Make default safety of align-content on block containers that are scroll containers unsafe.

fantasai opened the pull request:
#### fd49ec9a304465a0e383ecd001823c9cc354cb38
<pre>
Make default safety of align-content on block containers that are scroll containers unsafe.
<a href="https://bugs.webkit.org/show_bug.cgi?id=272002">https://bugs.webkit.org/show_bug.cgi?id=272002</a>
<a href="https://rdar.apple.com/problem/125742095">rdar://problem/125742095</a>

Reviewed by Simon Fraser.

Tweaks the align-content safety checks for RenderBlockFlow to assume
&quot;unsafe&quot; rather than &quot;safe&quot; alignment on scroll containers.

* LayoutTests/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-block-overflow-000-expected.html: Adjust expectations.
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-block-overflow-000-ref.html: Adjust expectations.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-block-001.html: Adjust expectations.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::shiftForAlignContent): Check for &apos;safe&apos; and content distribution explicitly, limit default safety check to non-scroll-containers.
(WebCore::RenderBlockFlow::allowedLayoutOverflow const): Simplify short-circuit alignment check.

Canonical link: <a href="https://commits.webkit.org/276929@main">https://commits.webkit.org/276929@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/620daaf73933ad9532210d36ec5d81a80f47722e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46184 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25315 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48769 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48855 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42225 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48491 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29646 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22759 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37736 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46762 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22381 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39825 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18943 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19773 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40923 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4228 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42478 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41272 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50653 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21180 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17667 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44913 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22480 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/39982 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43822 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10226 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22839 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22174 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -678,7 +678,10 @@ inline LayoutUnit RenderBlockFlow::shiftForAlignContent(LayoutUnit intrinsicLogi
     LayoutUnit space = computedLogicalHeight - intrinsicLogicalHeight;
-    if (space <= 0 && OverflowAlignment::Unsafe != alignment.overflow())
+    if (space <= 0
+        && (OverflowAlignment::Safe == alignment.overflow()
```
We don't generally put the values on the LHS of an equality check (which people do to avoid assignment in the case of a typo), so I'd flip these around to match other code.

smfr replied:
Also, with these more complex boolean conditions, it often helps readability to use a lambda function with a helpful name, which I think would improve things in this case. It's hard to tell what this condition is trying to do.

fantasai replied:
Would something like this work?
```
    if (space <= 0) {
        bool overflowIsSafe = alignment.overflow() == OverflowAlignment::Default && !isScrollContainerY()
            || alignment.overflow() == OverflowAlignment::Safe
            || alignment.position() == ContentPosition::Normal;
        if (overflowIsSafe)
            return 0_lu; // Floored at zero; we're done
    }
```

It's a temp variable instead of a lambda, but I think that seems simpler than a lambda.

fantasai committed:
Make default safety of align-content on block containers that are scroll containers unsafe.
https://bugs.webkit.org/show_bug.cgi?id=272002
rdar://problem/125742095

Reviewed by Simon Fraser.

Tweaks the align-content safety checks for RenderBlockFlow to assume
"unsafe" rather than "safe" alignment on scroll containers.

* LayoutTests/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-block-overflow-000-expected.html: Adjust expectations.
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-block-overflow-000-ref.html: Adjust expectations.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-block-001.html: Adjust expectations.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::shiftForAlignContent): Check for 'safe' and content distribution explicitly, limit default safety check to non-scroll-containers.
(WebCore::RenderBlockFlow::allowedLayoutOverflow const): Simplify short-circuit alignment check.

Canonical link: https://commits.webkit.org/276929@main
